File: web/js/detection.js
/**
 * Sprint End Detection — BIDIRECTIONAL
 *
 * Port of detection.py for use in the web analyzer.
 *
 * FORWARD (from sprint phase):
 *   Find where acceleration rolling mean first drops significantly below
 *   the sustained sprint level.
 *
 * BACKWARD (from recording end):
 *   Work backward through standing/walking/decel to find where acceleration
 *   was still at sprint level.
 *
 * When both agree → high confidence.
 * When they disagree → the gap is the uncertainty zone.
 */

// Agreement threshold: if forward and backward are within this many seconds,
// they "agree" and we average them.
const AGREEMENT_THRESHOLD = 1.5;

const MIN_TIMES = { 60: 3, 70: 4, 100: 5, 200: 15, 290: 25, 400: 40 };

// =============================================================================
// Data loading
// =============================================================================

function decodeCurve(encodedData) {
  return JSON.parse(atob(encodedData));
}

function loadSprintData(sprint) {
  const result = {};
  if (sprint.accelerationCurve) {
    const raw = decodeCurve(sprint.accelerationCurve);
    const accel = raw
      .map((d) => ({
        timestamp: d.timestamp,
        accel_x: d.x,
        accel_y: d.y,
        accel_z: d.z,
      }))
      .sort((a, b) => a.timestamp - b.timestamp);
    for (const row of accel) {
      row.mag = Math.sqrt(
        row.accel_x ** 2 + row.accel_y ** 2 + row.accel_z ** 2
      );
    }
    result.accel = accel;
  }
  if (sprint.gyroscopeCurve) {
    const raw = decodeCurve(sprint.gyroscopeCurve);
    const gyro = raw
      .map((d) => ({
        timestamp: d.timestamp,
        gyro_x: d.x,
        gyro_y: d.y,
        gyro_z: d.z,
      }))
      .sort((a, b) => a.timestamp - b.timestamp);
    result.gyro = gyro;
  }
  return result;
}

function getSprintsWithCurves(data) {
  const sprints = [];
  for (const session of data.sessions) {
    const sessionDate = session.date.slice(0, 10);
    for (const sprint of session.sprints || []) {
      if (sprint.accelerationCurve && sprint.gyroscopeCurve) {
        sprints.push({
          date: sessionDate,
          distance: sprint.distance || 0,
          sprint,
        });
      }
    }
  }
  return sprints;
}

// =============================================================================
// Common: rolling mean + sprint level
// =============================================================================

function computeRollingMean(accel, windowSeconds = 1.0) {
  const n = accel.length;
  const t0 = accel[0].timestamp;
  const tEnd = accel[n - 1].timestamp;
  const rate = n / (tEnd - t0);
  const window = Math.max(10, Math.round(windowSeconds * rate));
  const minPeriods = 5;

  const mag = accel.map((d) => d.mag);
  const rolling = new Array(n);

  for (let i = 0; i < n; i++) {
    const start = Math.max(0, i - window + 1);
    const count = i - start + 1;
    if (count < minPeriods) {
      rolling[i] = null;
    } else {
      let sum = 0;
      for (let j = start; j <= i; j++) {
        sum += mag[j];
      }
      rolling[i] = sum / count;
    }
  }

  return { rolling, rate };
}

function findSprintLevel(timestamps, rolling) {
  const t0 = timestamps[0];
  const duration = timestamps[timestamps.length - 1] - t0;
  const midStart = t0 + duration * 0.2;
  const midEnd = t0 + duration * 0.7;

  const midValues = [];
  for (let i = 0; i < timestamps.length; i++) {
    if (
      timestamps[i] >= midStart &&
      timestamps[i] <= midEnd &&
      rolling[i] !== null
    ) {
      midValues.push(rolling[i]);
    }
  }

  if (midValues.length < 10) return 5.0;
  return median(midValues);
}

function findSprintStart(timestamps, rolling) {
  for (let i = 0; i < rolling.length; i++) {
    if (rolling[i] !== null && rolling[i] > 2.0) {
      return { time: timestamps[i], index: i };
    }
  }
  return { time: timestamps[0], index: 0 };
}

// =============================================================================
// FORWARD detection
// =============================================================================

function detectForward(timestamps, rolling, sprintLevel, minSprintTime = 5.0) {
  const threshold = sprintLevel * 0.9;
  const searchStart = timestamps[0] + minSprintTime;
  const n = timestamps.length;
  const rate = n / (timestamps[n - 1] - timestamps[0]);
  const sustainSamples = Math.max(5, Math.round(0.5 * rate));

  for (let i = 0; i < n; i++) {
    if (timestamps[i] < searchStart) continue;
    if (rolling[i] === null) continue;
    if (rolling[i] < threshold) {
      let sustained = true;
      const endCheck = Math.min(i + sustainSamples, n);
      for (let j = i; j < endCheck; j++) {
        if (rolling[j] !== null && rolling[j] >= threshold) {
          sustained = false;
          break;
        }
      }
      if (sustained) {
        return { time: timestamps[i], threshold };
      }
    }
  }

  return { time: timestamps[n - 1], threshold };
}

// =============================================================================
// BACKWARD detection
// =============================================================================

function detectBackward(timestamps, rolling, sprintLevel) {
  const threshold = sprintLevel * 0.9;
  for (let i = timestamps.length - 1; i >= 0; i--) {
    if (rolling[i] === null) continue;
    if (rolling[i] >= threshold) {
      return { time: timestamps[i], threshold };
    }
  }
  return { time: timestamps[0], threshold };
}

// =============================================================================
// DECISION: combine forward + backward
// =============================================================================

function decide(forwardTime, backwardTime, _sprintStart) {
  const gap = Math.abs(forwardTime - backwardTime);
  if (gap <= AGREEMENT_THRESHOLD) {
    const final = (forwardTime + backwardTime) / 2;
    return { final, decision: 'agree', gap };
  }
  if (forwardTime < backwardTime) {
    return { final: backwardTime, decision: 'trust_backward', gap };
  }
  return { final: forwardTime, decision: 'trust_forward', gap };
}

// =============================================================================
// Utilities
// =============================================================================

function median(arr) {
  const sorted = [...arr].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  if (sorted.length % 2 === 0) {
    return (sorted[mid - 1] + sorted[mid]) / 2;
  }
  return sorted[mid];
}

function decodeSplits(encoded) {
  if (!encoded) return [];
  try {
    return JSON.parse(atob(encoded));
  } catch {
    return [];
  }
}

function variance(arr) {
  const m = arr.reduce((a, b) => a + b, 0) / arr.length;
  return arr.reduce((s, v) => s + (v - m) ** 2, 0) / arr.length;
}

// =============================================================================
// High-level: analyze file data
// =============================================================================

export function analyzeData(data) {
  const sprints = getSprintsWithCurves(data).filter((s) => s.distance >= 60);

  const results = [];
  for (let i = 0; i < sprints.length; i++) {
    const sprintInfo = sprints[i];
    const sprint = sprintInfo.sprint;
    const dist = sprintInfo.distance;
    const date = sprintInfo.date;

    const loaded = loadSprintData(sprint);
    const accel = loaded.accel;
    if (!accel || accel.length < 100) continue;

    const timestamps = accel.map((d) => d.timestamp);
    const { rolling, rate } = computeRollingMean(accel);
    const sprintLevel = findSprintLevel(timestamps, rolling);
    const { time: sprintStart } = findSprintStart(timestamps, rolling);

    const minTime = MIN_TIMES[dist] ?? 5;

    const fwd = detectForward(timestamps, rolling, sprintLevel, minTime);
    const bwd = detectBackward(timestamps, rolling, sprintLevel);
    const { final: finalTime, decision, gap } = decide(
      fwd.time,
      bwd.time,
      sprintStart
    );

    const fwdDur = fwd.time - sprintStart;
    const bwdDur = bwd.time - sprintStart;
    const finalDur = finalTime - sprintStart;

    // Relative-time arrays for plotting
    const tRel = timestamps.map((t) => t - sprintStart);
    const rawMag = accel.map((d) => d.mag);

    // Gyroscope data
    const gyro = loaded.gyro;
    let gyroData = null;
    if (gyro && gyro.length > 100) {
      const gT = gyro.map((d) => d.timestamp - sprintStart);
      const gX = gyro.map((d) => d.gyro_x);
      const gY = gyro.map((d) => d.gyro_y);
      const gZ = gyro.map((d) => d.gyro_z);
      const vx = variance(gX), vy = variance(gY), vz = variance(gZ);
      const dominantAxis = vx >= vy && vx >= vz ? 'x' : vy >= vz ? 'y' : 'z';
      gyroData = { t: gT, x: gX, y: gY, z: gZ, dominantAxis };
    }

    // Sprint metadata from backup
    const meta = {
      reactionTime: sprint.reactionTime ?? null,
      isFalseStart: sprint.isFalseStart ?? false,
      peakPropulsiveG: sprint.peakPropulsiveG ?? null,
      avgPropulsiveG: sprint.avgPropulsiveG ?? null,
      maxGForce: sprint.maxGForce ?? null,
      avgCadence: sprint.avgCadence ?? null,
      stepCount: sprint.stepCount ?? null,
      avgStrideLength: sprint.avgStrideLength ?? null,
      maxVelocity: sprint.maxVelocity ?? null,
      timeToMaxVelocity: sprint.timeToMaxVelocity ?? null,
      armDriveFocus: sprint.armDriveFocus ?? null,
      peakArmVelocity: sprint.peakArmVelocity ?? null,
      gpsStatus: sprint.gpsVerificationStatus ?? 'notApplicable',
      splits: decodeSplits(sprint.splitTimesData),
    };

    results.push({
      index: i + 1,
      date,
      distance: dist,
      fwdDur,
      bwdDur,
      finalDur,
      gap,
      decision,
      meta,
      plotData: {
        t: tRel,
        rolling,
        rawMag,
        sprintLevel,
        threshold: fwd.threshold,
        fwdTime: fwdDur,
        bwdTime: bwdDur,
        finalTime: finalDur,
      },
      gyroData,
    });
  }

  return results;
}

File: detection.py
"""
Sprint End Detection — BIDIRECTIONAL

Extracted from DetectFromEnd.py for use in the TUI analyzer.

FORWARD (from sprint phase):
  Find where acceleration rolling mean first drops significantly below
  the sustained sprint level.

BACKWARD (from recording end):
  Work backward through standing/walking/decel to find where acceleration
  was still at sprint level.

When both agree → high confidence.
When they disagree → the gap is the uncertainty zone.
"""

import json
import base64
import numpy as np
import pandas as pd
from pathlib import Path

# Agreement threshold: if forward and backward are within this many seconds,
# they "agree" and we average them.
AGREEMENT_THRESHOLD = 1.5  # seconds

MIN_TIMES = {60: 3, 70: 4, 100: 5, 200: 15, 290: 25, 400: 40}


# =============================================================================
# Data loading
# =============================================================================

def decode_curve(encoded_data):
    return json.loads(base64.b64decode(encoded_data))


def load_sprint_data(sprint):
    result = {}
    if 'accelerationCurve' in sprint:
        accel = pd.DataFrame(decode_curve(sprint['accelerationCurve']))
        accel = accel.rename(columns={'x': 'accel_x', 'y': 'accel_y', 'z': 'accel_z'})
        accel = accel.sort_values('timestamp').reset_index(drop=True)
        accel['mag'] = np.sqrt(accel['accel_x']**2 + accel['accel_y']**2 + accel['accel_z']**2)
        result['accel'] = accel
    if 'gyroscopeCurve' in sprint:
        gyro = pd.DataFrame(decode_curve(sprint['gyroscopeCurve']))
        gyro = gyro.rename(columns={'x': 'gyro_x', 'y': 'gyro_y', 'z': 'gyro_z'})
        gyro = gyro.sort_values('timestamp').reset_index(drop=True)
        result['gyro'] = gyro
    return result


def get_sprints_with_curves(data):
    sprints = []
    for session in data['sessions']:
        session_date = session['date'][:10]
        for sprint in session.get('sprints', []):
            if sprint.get('accelerationCurve') and sprint.get('gyroscopeCurve'):
                sprints.append({
                    'date': session_date,
                    'distance': sprint.get('distance', 0),
                    'sprint': sprint,
                })
    return sprints


# =============================================================================
# Common: rolling mean + sprint level
# =============================================================================

def compute_rolling_mean(accel_df, window_seconds=1.0):
    t = accel_df['timestamp'].values
    mag = accel_df['mag'].values
    rate = len(accel_df) / (t[-1] - t[0])
    window = max(10, int(window_seconds * rate))
    rolling = pd.Series(mag).rolling(window=window, min_periods=5).mean().values
    return rolling, rate


def find_sprint_level(t, rolling):
    duration = t[-1] - t[0]
    mid_start = t[0] + duration * 0.2
    mid_end = t[0] + duration * 0.7
    mid_mask = (t >= mid_start) & (t <= mid_end) & ~np.isnan(rolling)
    if mid_mask.sum() < 10:
        return 5.0
    return np.median(rolling[mid_mask])


def find_sprint_start(t, rolling):
    for i in range(len(rolling)):
        if not np.isnan(rolling[i]) and rolling[i] > 2.0:
            return t[i], i
    return t[0], 0


# =============================================================================
# FORWARD detection
# =============================================================================

def detect_forward(t, rolling, sprint_level, min_sprint_time=5.0):
    threshold = sprint_level * 0.90
    search_start = t[0] + min_sprint_time
    rate = len(t) / (t[-1] - t[0])
    sustain_samples = max(5, int(0.5 * rate))

    for i in range(len(t)):
        if t[i] < search_start:
            continue
        if np.isnan(rolling[i]):
            continue
        if rolling[i] < threshold:
            sustained = True
            end_check = min(i + sustain_samples, len(t))
            for j in range(i, end_check):
                if not np.isnan(rolling[j]) and rolling[j] >= threshold:
                    sustained = False
                    break
            if sustained:
                return t[i], threshold

    return t[-1], threshold


# =============================================================================
# BACKWARD detection
# =============================================================================

def detect_backward(t, rolling, sprint_level):
    threshold = sprint_level * 0.90
    for i in range(len(t) - 1, -1, -1):
        if np.isnan(rolling[i]):
            continue
        if rolling[i] >= threshold:
            return t[i], threshold
    return t[0], threshold


# =============================================================================
# DECISION: combine forward + backward
# =============================================================================

def decide(forward_time, backward_time, sprint_start):
    gap = abs(forward_time - backward_time)
    if gap <= AGREEMENT_THRESHOLD:
        final = (forward_time + backward_time) / 2
        return final, 'agree', gap
    if forward_time < backward_time:
        return backward_time, 'trust_backward', gap
    return forward_time, 'trust_forward', gap


# =============================================================================
# High-level: analyze a file
# =============================================================================

def _analyze_sprint(accel_df, distance, date, index):
    """Run bidirectional detection on a single sprint's acceleration DataFrame.

    Returns a result dict or None if insufficient data.
    """
    if accel_df is None or len(accel_df) < 100:
        return None

    t = accel_df['timestamp'].values
    rolling, rate = compute_rolling_mean(accel_df)
    sprint_level = find_sprint_level(t, rolling)
    sprint_start, _ = find_sprint_start(t, rolling)

    min_time = MIN_TIMES.get(distance, 5)

    fwd_time, fwd_thresh = detect_forward(t, rolling, sprint_level, min_sprint_time=min_time)
    bwd_time, _ = detect_backward(t, rolling, sprint_level)
    final_time, decision, gap = decide(fwd_time, bwd_time, sprint_start)

    fwd_dur = fwd_time - sprint_start
    bwd_dur = bwd_time - sprint_start
    final_dur = final_time - sprint_start

    return {
        'index': index,
        'date': date,
        'distance': distance,
        'fwd_dur': fwd_dur,
        'bwd_dur': bwd_dur,
        'final_dur': final_dur,
        'gap': gap,
        'decision': decision,
        'plot_data': {
            't': t - sprint_start,
            'rolling': rolling,
            'sprint_level': sprint_level,
            'threshold': fwd_thresh,
            'fwd_time': fwd_dur,
            'bwd_time': bwd_dur,
            'final_time': final_dur,
        },
    }


def analyze_file(path):
    """
    Analyze a .falcata file. Returns a list of result dicts, each containing:
      - index, date, distance, fwd_dur, bwd_dur, final_dur, gap, decision
      - plot_data: dict with t, rolling, sprint_level, threshold, sprint_start,
                   fwd_time, bwd_time, final_time (for charting)
    """
    with open(path, 'r') as f:
        data = json.load(f)

    sprints = get_sprints_with_curves(data)
    sprints = [s for s in sprints if s['distance'] >= 60]

    results = []
    for i, sprint_info in enumerate(sprints):
        sprint = sprint_info['sprint']
        loaded = load_sprint_data(sprint)
        r = _analyze_sprint(loaded.get('accel'), sprint_info['distance'], sprint_info['date'], i + 1)
        if r is not None:
            results.append(r)

    return results


def analyze_firestore_sessions(sessions):
    """Analyze sprints from Firestore session documents.

    Args:
        sessions: list of Firestore session dicts (from firestore_loader.fetch_sessions).

    Returns same format as analyze_file().
    """
    from firestore_loader import session_to_sprints

    results = []
    idx = 1
    for session in sessions:
        for sprint_info in session_to_sprints(session):
            if sprint_info['distance'] < 60:
                continue
            accel_df = sprint_info['loaded'].get('accel')
            r = _analyze_sprint(accel_df, sprint_info['distance'], sprint_info['date'], idx)
            if r is not None:
                results.append(r)
                idx += 1

    return results
